src/datadog/grok/filters/keyvalue.rs

```diff
@@ -302,84 +268,57 @@ fn quoted<'a>(
-            )))
-        }
-    }
-}
-
-/// Parses an input while it matches a given regex, otherwise skips an input until the next field delimiter
-#[inline]
-fn match_re_or_empty<'a>(
-    value_re: &'a Regex,
-    field_delimiter: &'a str,
-) -> impl Fn(&'a str) -> SResult<&'a str> {
-    move |input| {
-        re_find::<'a, (&'a str, nom::error::ErrorKind)>(value_re)(input)
-            .or_else(|_| parse_delimited(field_delimiter)(input).map(|(rest, _v)| (rest, "")))
-    }
-}
-
 /// Parses the value.
 /// The value has two parsing strategies.
 ///
 /// 1. The value is quoted - parse until the end quote
-/// 2. Otherwise we parse until regex matches
+/// 2. Otherwise, we parse until regex matches
 #[inline]
-fn parse_value<'a>(
-    field_delimiter: &'a str,
-    quotes: &'a [(char, char)],
-    re: &'a Regex,
-) -> impl Fn(&'a str) -> SResult<Value> {
-    move |input| {
-        alt((
-            map(quoted(quotes, field_delimiter), |value| {
-                Value::Bytes(Bytes::copy_from_slice(value.as_bytes()))
-            }),
-            parse_null,
-            parse_boolean,
-            parse_number(field_delimiter),
-            map(match_re_or_empty(re, field_delimiter), |value| {
-                Value::Bytes(Bytes::copy_from_slice(value.as_bytes()))
-            }),
-        ))(input)
-    }
+fn parse_value<'a>(input: &'a str, quotes: &'a [(char, char)]) -> SResult<'a, Value> {
+    alt((
+        parse_null,
+        parse_boolean,
+        parse_number,
+        quoted(quotes).and_then(parse_string),
+        parse_string,
+    ))(input)
 }
 
-fn parse_number(field_delimiter: &str) -> impl Fn(&str) -> SResult<Value> + '_ {
-    move |input| {
-        map(
-            terminated(
-                double,
-                peek(alt((
-                    parse_field_delimiter(field_delimiter),
-                    parse_end_of_input(),
-                ))),
-            ),
-            |v| {
-                if ((v as i64) as f64 - v).abs() == 0.0 {
-                    // can be safely converted to Integer without precision loss
-                    Value::Integer(v as i64)
-                } else {
-                    Value::Float(NotNan::new(v).expect("not a float"))
-                }
-            },
-        )(input)
-        .map_err(|e| match e {
-            // double might return Failure(an unrecoverable error) - make it recoverable
-            nom::Err::Failure(_) => nom::Err::Error((input, nom::error::ErrorKind::Float)),
-            e => e,
-        })
+fn parse_string(input: &str) -> SResult<Value> {
+    map(rest, |s: &str| {
+        Value::Bytes(Bytes::copy_from_slice(s.trim().as_bytes()))
+    })(input)
+}
+
+fn parse_number(input: &str) -> SResult<Value> {
+    let res = map(terminated(double, eof), |v| {
+        if ((v as i64) as f64 - v).abs() == 0.0 {
+            // can be safely converted to Integer without precision loss
+            Value::Integer(v as i64)
```

Comment:
I was thinking to move the `validate_octal_number` bits right here so you don't have to re-match on the result value being an `Integer` since you already know it is. I see that doesn't work with the `map` function, is there another `nom` function that would allow returning an error?